compiler/qsc_frontend/src/compile/preprocess.rs

```diff
@@ -118,23 +118,27 @@ impl MutVisitor for Conditional {
     }
 }
 
-fn matches_target(attrs: &[Box<Attr>], target: TargetProfile) -> bool {
+fn matches_config(attrs: &[Box<Attr>], capabilities: RuntimeCapabilityFlags) -> bool {
     attrs.iter().all(|attr| {
         if hir::Attr::from_str(attr.name.name.as_ref()) == Ok(hir::Attr::Config) {
             if let ExprKind::Paren(inner) = attr.arg.kind.as_ref() {
                 match inner.kind.as_ref() {
-                    ExprKind::Path(path) => {
-                        match TargetProfile::from_str(path.name.name.as_ref()) {
-                            Ok(t) => t == target,
-                            Err(()) => true,
-                        }
-                    }
-                    _ => true,
+                    // If there is no config attribute, then we assume that the item matches
+                    // the target. We can't do membership tests on the capabilities because
+                    // Base is not a subset of any capabilities, it is a lack of capabilities.
+                    ExprKind::Path(path) => match ConfigAttr::from_str(path.name.name.as_ref()) {
+                        Ok(ConfigAttr::Full) => capabilities.is_all(),
+                        Ok(ConfigAttr::Base) => capabilities.is_empty(),
+                        _ => true,
+                    },
+                    _ => true, // Unknown config attribute, so we assume it matches
```

Comment:
`matches_config` is currently used to check whether to include an item in the compilation. However, it can silently behave in an unexpected way, most likely including something that was not meant to be included (e.g. `@Config(Ful)` when compiling for `Base`).

It might be worth to not silently include things for which the `@Config` attribute is not "well-defined", but we could wait until we decide how the attribute will be used moving forward.

Reply:
This is the same behavior as the original code, just commented to explain what is going on. @swernli - do you want to weigh in here?

Reply:
This behavior is intentional. The code in lowering checks whether or not the `@Config` attribute is correct, but if we drop stuff we don't recognize then it won't get checked by lowering. So we keep things we don't recognize and let it trigger the appropriate failure in a later check.